pilot/pkg/serviceregistry/kube/controller/authorization.go

```diff
@@ -419,21 +419,28 @@ func (c *Controller) PeerAuthenticationHandler(old config.Config, obj config.Con
 
 func (c *Controller) calculateUpdatedWorkloads(pods map[string]*v1.Pod) map[model.ConfigKey]struct{} {
 	updates := map[model.ConfigKey]struct{}{}
-	for _, pod := range pods {
+
+	updateForPod := func(pod *v1.Pod) {
```

Comment:
Note to reviewers: This fixes an actual race condition bug that seemed to be exposed by the updated tests.

Reply:
extractWorkload conflicts with what operation?

Reply:
It directly accesses `ambientIndex.waypoints`. Quite a bit of the code is doing this, unfortunately. We should probably refactor the code to better protect the `ambientIndex` eventually. For now, this code is definitely a problem and was leading to test failures.

Reply:
FYI I opened https://github.com/istio/istio/issues/45816 to more broadly address the fragility of the ambient controller.